dvc/repo/add.py

```diff
@@ -14,20 +15,29 @@
 from ..utils import LARGE_DIR_SIZE, glob_targets, resolve_paths
 from . import locked
 
+if TYPE_CHECKING:
+    from dvc.types import TargetType
+
+
 logger = logging.getLogger(__name__)
 
 
 @locked
 @scm_context
 def add(
-    repo, targets, recursive=False, no_commit=False, fname=None, **kwargs,
+    repo,
+    targets: "TargetType",
```

Comment:
I'd have used `TargetType` and `ensure_list` in more places, but most of the places need to leave `None` as-is to be used in `collect` or `used_cache`, so we need something like `ensure_list_or_none()`. Let me know if that makes sense, I'll add them in the next PR.

```python
def ensure_list_or_none(item: Union[List[str], str, None]) -> Optional[List[str]]:
    if item is None:
        return item
    if isinstance(item, str):
       return [item]
    return list(item)
```

Reply:
I am talking about the following stuff that are repeated on most of the API.
https://github.com/iterative/dvc/blob/ccc17917cfcb1b9ab8bc40d5ffdc8bcd02a1040a/dvc/repo/push.py#L22-L23

Reply:
A bit ugly, but makes sense.